design/design-doc-developer-experience-tooling.md

```diff
@@ -0,0 +1,1146 @@
+create an image.
+
+Initially, all supported builders can be built as part of the DevEx tooling,
+ensuring that we provide an experience that works out of the box. In the future,
+we may expose builders as an extension point, allowing external builder
+implementations to be configured. The design of builders is inspired by [Cloud
+Native Buildpacks], which could themselves be used as a builder implementation.
+
+### XRDs, Compositions, and Operations
+
+XRDs, Compositions, and Operations in a project are regular Crossplane
+resources. To invoke an embedded function in a composition or operation
+pipeline, the user refers to it by package name, just like any other function.
+
+Note that the package name for an embedded function is constructed in the same
+manner used by the Crossplane package manager when resolving dependencies, based
+on the repository naming scheme described above.
+
+Example:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: xexample.example.org
+spec:
+  mode: Pipeline
+  pipeline:
+    - step: compose
+      functionRef:
+        name: examplecom-my-platformcompose-cluster
+    - step: propagate-status
+      functionRef:
+        name: examplecom-my-platformpropagate-status
+    - step: function-auto-ready
+      functionRef:
+        name: crossplane-contrib-function-auto-ready
+```
+
+The tooling will include helper commands for building XRDs, compositions, and
+operations. For example, we can generate an XRD from an example XR (inferring
+the OpenAPI spec), convert other API specification formats (e.g., kro's [Simple
+Schema]) to XRDs, scaffold a composition for an XRD, and add steps to pipelines.
+
+### Language Bindings (Schemas)
+
+To make it easier to build functions, we will provide tools to generate language
+bindings (also referred to as schemas) for XRDs and CRDs. The mechanism used to
+generate language bindings varies by language; current implementations in the
+Upbound tooling are:
+
+* **Go:** [oapi-codegen], with some custom mutation to produce better code.
+* **go-templating:** Direct conversion of OpenAPI specs to [JSONSchema], with
+  some custom mutation to produce better schemas.
+* **KCL:** `kcl import` to convert CRDs to KCL schemas.
+* **Python:** [datamodel-code-generator], producing [Pydantic] models.
+
+The generated language bindings provide two advantages when authoring functions:
+
+1. Type safety and schema validation, to varying degrees depending on the
+   language.
+2. IDE integration via standard tooling, providing features like autocomplete,
+   inline documentation, linting, etc.
+
+Language bindings can be generated for any Crossplane package that contains CRDs
+or XRDs. Generation can run client-side (as part of pulling a dependency package
+or manually via a command), but the bindings can also be packaged as part of a
+Crossplane package to avoid the need to generate them on download.
```

Comment:
FWIW I've found this delivery mechanism for bindings to be a little awkward. While working with Python I found myself wishing the bindings were a regular Python package (sdist, wheel, etc) that I could `pip install`.

Today they're generated into `.up/` at the project root and then we do some symlink hacks to make them load. I can get it working with a virtualenv but I don't find it very intuitive. I wonder if we'll find the pattern causes compatibility/flexibility issues down the line.